python/conversion.hpp

```diff
@@ -66,6 +66,17 @@ std::optional<T> py2optional(pybind11::object o, const char* msg) {
     return o.is_none()? std::nullopt: std::optional<T>(std::move(value));
 }
 
+template <typename T, typename F>
```

Comment:
```cxx
template <typename T, typename F>
std::optional<T> optional_when(T&& o, F&& pred) {
    if (pred(std::forward(T)) {
      return {std::forward(T)};
    } else {
      return {};
    }
}
```
- A slightly more descriptive name
- Do not construct the optional if not needed
- Play nicely with rvalue refs, lvalue ref, and values.